logsmanagement/circular_buffer.c

```diff
@@ -0,0 +1,405 @@
+ * Note that buff->tail can only be changed through circ_buff_read_done(), and 
+ * circ_buff_search() and circ_buff_read_done() are mutually exclusive due 
+ * to uv_mutex_lock() and uv_mutex_unlock() in queries and when writing to DB.
+ * 
+ * @param p_query_params Query parameters to search according to.
+ * @param p_file_infos   File_info structs to be searched.
+ */
+void circ_buff_search(logs_query_params_t *const p_query_params, struct File_info *const p_file_infos[]) {
+    
+    for(int pfi_off = 0; p_file_infos[pfi_off]; pfi_off++)
+        uv_rwlock_rdlock(&p_file_infos[pfi_off]->circ_buff->buff_realloc_rwlock);
+
+    int buffs_size = 0, 
+        buff_max_num_of_items = 0;
+
+    while(p_file_infos[buffs_size]){
+        if(p_file_infos[buffs_size]->circ_buff->num_of_items > buff_max_num_of_items) 
+            buff_max_num_of_items = p_file_infos[buffs_size]->circ_buff->num_of_items;
+        buffs_size++;
+    }
+
+    struct qsort_item items[buffs_size * buff_max_num_of_items + 1]; // worst case allocation
+    
+    int items_off = 0;
+
+    for(int buff_off = 0; p_file_infos[buff_off]; buff_off++){
+        Circ_buff_t *buff = p_file_infos[buff_off]->circ_buff;
+        /* TODO: The following 3 operations need to be replaced with a struct
+         * to gurantee atomicity. */
+        int head = __atomic_load_n(&buff->head, __ATOMIC_SEQ_CST) % buff->num_of_items;
+        int tail = __atomic_load_n(&buff->tail, __ATOMIC_SEQ_CST) % buff->num_of_items;
+        int full = __atomic_load_n(&buff->full, __ATOMIC_SEQ_CST);
+
+        if ((head == tail) && !full) continue;  // Nothing to do if buff is empty
+
+        for (int i = tail; i != head; i = (i + 1) % buff->num_of_items){
+            items[items_off].cbi = &buff->items[i];
+            items[items_off++].pfi = p_file_infos[buff_off];
+        }
+    }
+
+    items[items_off].cbi = NULL;
+    items[items_off].pfi = NULL;
+
+    if(items[0].cbi) 
+        qsort(items, items_off, sizeof(items[0]), p_query_params->order_by_asc ? qsort_timestamp : reverse_qsort_timestamp);
+
+    
+    BUFFER *const res_buff = p_query_params->results_buff;
+ 
+    logs_query_res_hdr_t res_hdr = { // result header
+        .timestamp = p_query_params->act_to_ts,
+        .text_size = 0,
+        .matches = 0,
+        .log_source = "",
+        .log_type = ""
+    }; 
+
+    for (int i = 0; items[i].cbi; i++) {
+
+        /* If exceeding quota or timeout is reached and new timestamp is different than previous, 
+         * terminate query but inform caller about act_to_ts to continue from (its next value) in next call. */
+        if((res_buff->len >= p_query_params->quota || now_monotonic_usec() > p_query_params->stop_monotonic_ut) && 
+                items[i].cbi->timestamp != res_hdr.timestamp){
+            p_query_params->act_to_ts = res_hdr.timestamp;
+            break;
+        }
+
+        res_hdr.timestamp = items[i].cbi->timestamp;
+        res_hdr.text_size = items[i].cbi->text_size;
```

Comment:
The function `snprinftf` has a slow algorithm when compared with `strcpy`, please, change calls here for `strncpyz`.